### ISSUE TITLE
Entity cant have multiple Connectors

### ISSUE THREAD

**Issue by vytsci** [oroinc/crm#317]
Hello, Im new here, so Im open to critics about posting issue properly. Ive came up with a problem where you need to import same entity from multiple sources, same entity should have different connectors, different data converters, but based on entity-config architecture entity only can have 1 connector and 1 integration that it belongs to. Is this an issue, or maybe there are other solutions to my problem? The only solution i found is to create entity as mapped superclass and then extend it as many times as I have sources.

**Comment by anyt** [comment]
Hi vytsci,

I don't get what is the issue. In channel_id you can store different channels. The difference between Integrations will be mostly in converters.

> based on entity-config architecture entity only can have 1 connector and 1 integration that it belongs to
Not exactly. Regular entity or channel subject can be used in different channels or integrations. As an example you can see B2bCustomer entity.

**Comment by vytsci** [comment]
Customer entities are bad example, because they are used differently than other entities. It seems you you have a bundle where you have an entity named for example Order if you try to import your orders from multiple integrations, for example accounting software and frontend store, you will end up having a very bad time doing so. Because ORO attached only last channel it finds to that entity. And if you have for example 3 different channels and 3 different integrations ORO will attach only last one to that entity. Technically you can workaround this by using same names for each channel or integration, but if i want to make for example pragma_order_channel and prestashop_order_channel, my pragma channel wont work, because oro will attach prestashop_order_channel to my entity and when processing pragma_order_channel will have no attachment to my entity. Entity will see that integration and channel exists, but wont perform any tasks on my entity, because entity has no relation with channel it has not been attached to. If i make channel names exactly the same i.e. order_channel then everything works. But there may be cases where this wont work.

**Comment by vytsci** [comment]
Maybe i configured something badly, but my entities have only single channel name attached to them via entity config. Everything above entity config part can support multiple channels and integrations.
`yml
      channels:
          entity_data:
             -
                name: Oro\Bundle\SomeEntity\Entity\RealEntity                # Entity FQCN
                dependent:                                                      # Service entities that dependent on availability of main entity
                      - Oro\Bundle\SomeEntity\Entity\RealEntityStatus
                      - Oro\Bundle\SomeEntity\Entity\RealEntityCloseReason
                navigation_items:                                               # Navigation items that responsible for entity visibility
                      - menu.tab.real_entity_list

             -
                name: Oro\Bundle\AcmeDemoBundle\Entity\AnotherEntity
                dependent: ~
                navigation_items:
                    - menu.tab.entity_funnel_list
                    - menu.tab.some_tab.some_tab.some_value
                belongs_to:
                    integration: integration_type_name                   # If entity belongs to integration, correspondent node should be set
                    connector:   another                                 # connector name`
If I want to have multiple connectors and multiple integrations to import one entity, technically I will have problems doing so. Actually integration type must be the same for all integrations i.e. if I have "pragma_integration_type" and "prestashop_integration_type", the only solution is to name then "integration_type" I need to skip unique prefix for both integrations to work.

**Comment by dxops** [comment]
\Oro\Bundle\IntegrationBundle\Entity\Channel: you can process the same entity using multiple Integrations

\Oro\Bundle\ChannelBundle\Entity\Channel: there are no multiple integrations per channel support,
so options are:
1) use \Oro\Bundle\IntegrationBundle\Entity\Channel and use FeatureToggleBundle to disable menus and routes
2) customize \Oro\Bundle\ChannelBundle\Entity\Channel
2a) \Oro\Bundle\ChannelBundle\DependencyInjection\ChannelConfiguration::getConfigTreeBuilder
update config to support multiple integrations
\Oro\Bundle\ChannelBundle\Provider\SettingsProvider::getIntegrationConnectorName use updated config with multiple integrations
2b) Customize listener and return whatever you need \Oro\Bundle\MagentoBundle\EventListener\UpdateIntegrationConnectorsListener::getConnectors и \Oro\Bundle\ChannelBundle\EventListener\UpdateIntegrationConnectorsListener::getConnectors

**Comment by vytsci** [comment]
Yes Ive came up with solution, Ive used same names for integrations and channels. This works as workaround, but in some cases this may be a very hard to find workaround.
